core/src/main/scala/org/apache/spark/SparkContext.scala

```diff
@@ -1627,7 +1627,10 @@ class SparkContext(config: SparkConf) extends Logging with ExecutorAllocationCli
       partitions: Seq[Int],
       allowLocal: Boolean
       ): Array[U] = {
-    runJob(rdd, (context: TaskContext, iter: Iterator[T]) => func(iter), partitions, allowLocal)
+    // We must clean `func` here before using it in another closure below
+    // Otherwise, the closure cleaner will only clean the outer closure but not `func`
+    val cleanedFunc = clean(func)
```

Comment:
Ok, I have decided to fix this in a separate patch since it's really a separate issue. I have filed [SPARK-7237](https://issues.apache.org/jira/browse/SPARK-7237) for that